File: 36. Valid Sudoku.rb
#  https://leetcode.com/problems/valid-sudoku/

# @param {Character[][]} board
# @return {Boolean}
def is_valid_sudoku(board)
  valid_row?(board) && valid_column?(board) && valid_all_trio?(board)
end

def valid_row?(board)
  9.times do |i|
      map = {}
      
      9.times do |j|
          next if board[i][j] == '.'
          
          return false  if map[board[i][j]]
          map[board[i][j]] = true
      end
  end
  
  true
end

def valid_column?(board)
  9.times do |j|
      map = {}
      
      9.times do |i|
          next if board[i][j] == '.'
          
          return false  if map[board[i][j]]
          map[board[i][j]] = true
      end
  end
  
  true
end

def valid_trio?(board, s_row, s_col)
  map = {}
  3.times do |i|
      3.times do |j|
         next if board[s_row + i][s_col + j] == '.'
          
          return false if map[board[s_row + i][s_col + j]]
          map[board[s_row + i][s_col + j]] = true
      end
  end
  
  true
end

def valid_all_trio?(board)
  row = 0
  3.times do |i|
      col = 0
      3.times do |j|
          return false unless valid_trio?(board, row, col)
          col += 3
      end
      
      row += 3
  end
  
  true
end

File: leetcode-solutions/python/36. Valid Sudoku.py
# https://leetcode.com/problems/valid-sudoku/

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        sudoku = {
            'r0': {}, 'r1': {}, 'r2': {}, 'r3': {}, 'r4': {}, 'r5': {}, 'r6': {}, 'r7': {}, 'r8': {},
            'c0': {}, 'c1': {}, 'c2': {}, 'c3': {}, 'c4': {}, 'c5': {}, 'c6': {}, 'c7': {}, 'c8': {},
            's00': {}, 's01': {}, 's02': {}, 's10': {}, 's11': {}, 's12': {}, 's20': {}, 's21': {}, 's22': {},
        }

        for i in range(len(board)):
            for j in range(len(board[0])):
                r = 'r' + str(i)
                c = 'c' + str(j)
                s = 's' + str(i // 3) + str(j // 3)
                number = board[i][j]
                if number != '.':
                    if number in sudoku[r] or number in sudoku[c] or number in sudoku[s]:
                        print(sudoku)
                        return False
                    sudoku[r][number] = True
                    sudoku[c][number] = True
                    sudoku[s][number] = True
        return True


# shorter answer
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = [set() for _ in range(9)]
        col = [set() for _ in range(9)]
        sub = [set() for _ in range(9)]

        for i in range(len(board)):
            for j in range(len(board[0])):
                char = board[i][j]
                if char != '.':
                    sub_idx = (i // 3 * 3) + (j // 3)
                    if char in row[i] or char in col[j] or char in sub[sub_idx]:
                        return False
                    row[i].add(char)
                    col[j].add(char)
                    sub[sub_idx].add(char)
        return True
